src/goal.rs

```diff
@@ -1039,15 +926,68 @@ impl<'a> Goal<'a> {
   fn instantiate_constructor(con_ty: &Type, actual: &Type) -> Vec<Type> {
     let (args, ret) = con_ty.args_ret();
     let instantiations = find_instantiations(&ret, actual);
-    // println!("args: {:?}, ret: {}, actual: {:?}", args, ret, actual);
-    // println!("instantiations: {:?}", instantiations);
     let ret = args
       .iter()
       .map(|arg| Type::new(resolve_sexp(&arg.repr, &instantiations)))
       .collect();
-    // println!("new types: {:?}", ret);
     ret
   }
+
+  /// Add new grounding instantiations
+  /// that replace parent with child in previous instantiations
+  fn add_grounding(&mut self, parent: Symbol, child: Symbol) {
+    // First gather all the terms we want to instantiate:
+    // take both sides of the equation and all the premises
+    let mut sides = vec![&self.eq.lhs, &self.eq.rhs];
+    for premise in self.premises.iter() {
+      sides.push(&premise.lhs);
+      sides.push(&premise.rhs);
+    }
+
+    // Now create new instantiations from existing ones
+    let mut new_instantiations = vec![];
+    for inst in self.grounding_instantiations.iter() {
+      let replaced_canonicals: Vec<(Symbol, ETerm, bool)> = self
+        .params
+        .iter()
+        .map(|x| {
+          // Which class was this param instantiated to?
+          let id = inst.get(x).unwrap();
+          // Parameters must be canonical (at least in a clean state)
+          let canonical = CanonicalFormAnalysis::extract_canonical(&self.egraph, *id).unwrap();
+          // Try replacing the case-split variable with its child
+          let (new_expr, replaced) = replace_var(&canonical, parent, child);
+          let eterm = if replaced {
+            ETerm::new_from_expr(&new_expr, &mut self.egraph)
```

Comment:
**comment**: Why can't we do the instantiation here?

Reply:
You mean, as opposed to line 975? It's because to do the instantiation we need to have the whole substitution, whereas here we are dealing with each of its variables individually.

For example, say we have variables `n` and `xs` in the prop (IH), and `n` had already been previously split as `S n1` and `xs` as `Cons xs2 xs3`, and now we are splitting `n1` again into `S n4`. So we called `add_grounding` with `parent = n1` and `child = n4`. Then this loop will iterate twice (for each IH parameter) and the values could be as follows:
- First iteration: `x = n`, `id = 0` (the eclass of `n`), `canonical = S n1`, `(new_expr, replaced) = (S n4, true)`, `eterm` will create a new term `S n4` in the e-graph (this is only needed because we want its eclass id to store in the new instantiation, otherwise it would have also been added later, when we actually add the instantiated IH)
- Second iteration: `x = xs`, `id = 1` (the eclass of `xs`), `canonical = Cons xs2 xs3`,  `(new_expr, replaced) = (Cons xs2 xs3, false)` (because there was no `n1` in `canonical`), and the `eterm` will be `Cons xs2 xs3`, which is already in the e-graph.

Now that we have assembled the whole subst, which is `[n -> S n4, xs -> Cons xs2 xs3]`, we can actually do the instantiation after the loop (line 970 and after).